File: qyMessengerApp/qyMessengerApp/public/setFlg_forceKeyFrame.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace qy
{
    partial class qyFuncs
{
		public static unsafe int setFlg_forceKeyFrame(CCtxQmc pProcInfo, int iIndex_sharedObj)
		{
			int iErr = -1;

			bool bDbg = false;
			string str;

			//  MC_VAR_isCli	*	pProcInfo		=	QY_GET_procInfo_isCli(  );
			if (null==pProcInfo) return -1;

			QY_SHARED_OBJ pSharedObj = null;
			CAP_procInfo_bmpCommon pCapBmp = null;
			//TCHAR tBuf[128] = _T("");


			pSharedObj = getSharedObjByIndex(pProcInfo, iIndex_sharedObj);
			if (null==pSharedObj) goto errLabel;
			pCapBmp = getCapBmpBySth(pProcInfo, pSharedObj.var.iIndex_capBmp, 0);
			if (null==pCapBmp) goto errLabel;


			//
			uint dwTickCnt = myGetTickCount(null);

			//
			int iElapseInMs = (int)(dwTickCnt - pCapBmp.compressVideo.encInfo.dwTickCnt_startToForceKeyFrame);

			if (Math.Abs(iElapseInMs) > 500)
			{
				//	
				pCapBmp.compressVideo.encInfo.dwTickCnt_startToForceKeyFrame = dwTickCnt;

				//  2017/10/14
				str=string.Format("{0}, iIndex_sharedObj {1}", _T("setFlg_forceKeyFrame ok"), iIndex_sharedObj);
				showInfo_open(0, null, str);

			}


			iErr = 0;

		errLabel:

			if (0!=iErr)
			{
				str=string.Format("%s, iIndex_sharedObj %d", _T("setFlg_forceKeyFrame failed"), iIndex_sharedObj);
				showInfo_open(0, null, str);
			}

			return iErr;

		}

	}
}

File: srcProjs/srcRoot/src6/qyMcMain_isCli_common/qyMcMainPublic_isCli/public/setFlg_forceKeyFrame.cpp


#include	"stdafx.h"
#include	<time.h>
#include	<tchar.h>


#include	"qyMcMainCommon.h"
#include	"qmcVideoCapture.h"

//#include	"qmcVideoCapture_isCli.h"
//
#include	"qyDynLib.h"
#include	"qmcDmoPublic.h"
#include	"qmcCmdProc.h"
#include	"tmpCeLib.h"
#include	"module_qisCamCap.h"
#include	"isCmdConst.h"
#include	"qyOpenShellCommon.h"
#include	"qyCusResTemp.h"
#include	"policyAvParams.h"

#include	"qmcTaskPublic.h"

#include	"qmcCfg.h"


//
 __declspec(dllexport)  int  doAll_setFlg_forceKeyFrame(CCtxQmc* pProcInfo)
{
	//
	showInfo_open(0, 0, 0, _T("doAll_setFlg_forceKeyFrame called"));

	//
	for (int index = 0; index < pProcInfo->cfg.usMaxCnt_pSharedObjs; index++) {
		QY_SHARED_OBJ* pSharedObj = getSharedObjByIndex(pProcInfo, index);

		if (!pSharedObj->bUsed)  continue;

		//
		setFlg_forceKeyFrame(pProcInfo, index);
	}

	//
	return  0;
}




//  2014/08/25
__declspec(  dllexport  )  int  setFlg_forceKeyFrame(  MC_VAR_common  *  pProcInfo,  int  iIndex_sharedObj  )
 {
	 int				iErr			=  -1;
	 //  MC_VAR_isCli	*	pProcInfo		=	QY_GET_procInfo_isCli(  );
	 if  (  !pProcInfo  )  return  -1;

	 QY_SHARED_OBJ	*	pSharedObj		=  NULL;
	 CAP_procInfo_bmpU			*	pCapBmp					=	NULL;
	 TCHAR  tBuf[128]  =  _T(  ""  );


	 pSharedObj  =  getSharedObjByIndex(  pProcInfo,  iIndex_sharedObj  );	
	 if  (  !pSharedObj  )  goto  errLabel;
	 pCapBmp													=	getCapBmpBySth(  pProcInfo,  pSharedObj->iIndex_capBmp,  0  );
	if  (  !pCapBmp  )  goto  errLabel;

	
	//
	DWORD  dwTickCnt  =  GetTickCount(  );

	//
	int  iElapseInMs  =  dwTickCnt  -  pCapBmp->common.compressVideo.encInfo.dwTickCnt_startToForceKeyFrame;

	if  (  abs(  iElapseInMs  )  >  500  )  {
		//	
		pCapBmp->common.compressVideo.encInfo.dwTickCnt_startToForceKeyFrame  =  dwTickCnt;

		//  2017/10/14
		_sntprintf(  tBuf,  mycountof(  tBuf  ),  _T(  "%s, iIndex_sharedObj %d"  ),  _T(  "setFlg_forceKeyFrame ok"  ),  iIndex_sharedObj  );
		 showInfo_open0(  0,  0,  tBuf  );

	}
	

	 iErr  =  0;

errLabel:

	 if  (  iErr  )  {
		 _sntprintf(  tBuf,  mycountof(  tBuf  ),  _T(  "%s, iIndex_sharedObj %d"  ),  _T(  "setFlg_forceKeyFrame failed"  ),  iIndex_sharedObj  );
		 showInfo_open0(  0,  0,  tBuf  );
	 }

	 return  iErr;

 }


